### PR TITLE
refactor: use u16 representation for virtual node

### LINKED ISSUE #8281: Discussion: Extend vnode to u16
The vnode field is currently stored as a u8 in Hummock storage, representing 256 vnodes. However, the u8 field is now fully utilized, which means extending it in the future without breaking backward compatibility is impossible. We even haven't a reserved bit.

Given that 256 vnodes equates to 256 cores in the whole RisingWave cluster, it may be insufficient to support a medium one. Therefore, we suggest extending the field to u16, with the first byte always set to 0x00 now. This will allow for scalability in the future.

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Use `u16` as the internal representation and encoding format for virtual node. The total count of vnode is still 256 and the first byte is not used.

Resolve #8281 

Should we consider backward compatibility for this change?

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/hash/consistent_hash/vnode.rs [hunk 1]
```diff
@@ -20,14 +20,15 @@ use crate::hash::HashCode;
...
-/// `VirtualNode` (a.k.a. VNode) is a minimal partition that a set of keys belong to. It is used for
+/// `VirtualNode` (a.k.a. Vnode) is a minimal partition that a set of keys belong to. It is used for
 /// consistent hashing.
 #[repr(transparent)]
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Display)]
 #[display("{0}")]
 pub struct VirtualNode(VirtualNodeInner);
 
-type VirtualNodeInner = u8;
+/// The internal representation of a virtual node.
```
Comment (liurenjie1024) [inline src/common/src/hash/consistent_hash/vnode.rs:30]:
```suggestion
/// The internal representation of a virtual node id.
```

File: src/common/src/hash/consistent_hash/vnode.rs [hunk 2]
```diff
@@ -20,14 +20,15 @@ use crate::hash::HashCode;
...
+type VirtualNodeInner = u16;
```
Comment (liurenjie1024) [inline src/common/src/hash/consistent_hash/vnode.rs:31]: Would it be better to use newtype?

Comment (BugenZhao) [reply src/common/src/hash/consistent_hash/vnode.rs:31]: We'll keep this type private, and the exported type `VirtualNode` is the newtype. 🥺

**Comment by hzxa21** [comment]
> Should we consider backward compatibility for this change?

I think we should. But I am not sure how to do it since key encoding changes affect key comparison.

**Review by TennyZhuang — APPROVED** [review]

**Review by tabVersion — APPROVED** [review]

**Comment by hzxa21** [comment]
> > Should we consider backward compatibility for this change?
> 
> I think we should. But I am not sure how to do it since key encoding changes affect key comparison.

There are ways to ensure backward compatibility but they all looks hacky to me:
1. Let storage key comparator understand vnode instead of blinding comparing the user key. We can use ValueMeta to indicate how to decode vnode  (previsouly we have reserved a Byte in the value to store [flags](https://github.com/risingwavelabs/risingwave/blob/a37f5571caffadf0d48da7006d8e39934e429aa0/src/storage/src/hummock/value.rs#L63))
2. Similar to 1, but leverge SST format [version](https://github.com/risingwavelabs/risingwave/blob/a37f5571caffadf0d48da7006d8e39934e429aa0/src/storage/src/hummock/sstable/mod.rs#L269) to indicate how to decode vnode. This is doable since we also plan to bump SST format version in #8251 but it is more hacky than 1) because technically this is a user key encoding change, not a storage format change.
3. Similar to 1 and 2, but pass a custom user key comparator to storage to avoid letting storage understand vnode. Also seems to be hacky and overkill.

**Comment by hzxa21** [comment]
> Note that the vnode part of the key is assembled and also interpreted by the state table, and we may have no idea about whether there are one or two bytes in the storage...

Yes, this is the tricky part. If we choose to maintain backward compatibility, I believe storage need to handle the conversion. There are several hacky ways to let storage know wheter it is one or two bytes as I mentioned above.

**Comment by TennyZhuang** [comment]
Since we've broken several formats in the version, I guess the PR can be merged now.

**Pull request merged**